src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java

```diff
@@ -0,0 +1,259 @@
+        }
+        else if (learningObject instanceof LectureUnit lectureUnit) {
+            course = lectureUnit.getLecture().getCourse();
+        }
+        else {
+            throw new IllegalArgumentException("Learning object must be either LectureUnit or Exercise");
+        }
+        updateProgressByLearningObject(learningObject, new HashSet<>(userRepository.getStudents(course)));
+    }
+
+    /**
+     * Asynchronously update the existing progress for a specific learning goal
+     * @param learningGoal The learning goal for which to update all existing student progress
+     */
+    @Async
+    public void updateProgressByLearningGoalAsync(LearningGoal learningGoal) {
+        SecurityUtils.setAuthorizationObject(); // required for async
+        learningGoalProgressRepository.findAllByLearningGoalId(learningGoal.getId()).stream().map(LearningGoalProgress::getUser).forEach(user -> {
+            updateLearningGoalProgress(learningGoal.getId(), user);
+        });
+    }
+
+    /**
+     * Update the existing progress for a specific user in a course
+     * @param user The user for whom to update the existing learning goal progress
+     * @param course The course for which to fetch the learning goals from
+     * @return All learning goals of the course with the updated progress for the user
+     */
+    public Set<LearningGoal> getLearningGoalsAndUpdateProgressByUserInCourse(User user, Course course) {
+        var learningGoals = learningGoalRepository.findAllForCourse(course.getId());
+        learningGoals.forEach(learningGoal -> {
+            var updatedProgress = updateLearningGoalProgress(learningGoal.getId(), user);
+            if (updatedProgress != null) {
+                learningGoal.setUserProgress(Set.of(updatedProgress));
+            }
+        });
+        return learningGoals;
+    }
+
+    /**
+     * Update the progress for all learning goals linked to the given learning object
+     * @param learningObject The learning object for which to fetch the learning goals
+     * @param users A list of users for which to update the progress
+     */
+    public void updateProgressByLearningObject(ILearningObject learningObject, @NotNull Set<User> users) {
+        logger.debug("Updating learning goal progress for {} users.", users.size());
+        try {
+            Set<LearningGoal> learningGoals;
+            if (learningObject instanceof Exercise exercise) {
+                learningGoals = exerciseRepository.findByIdWithLearningGoals(exercise.getId()).map(Exercise::getLearningGoals).orElse(null);
+            }
+            else if (learningObject instanceof LectureUnit lectureUnit) {
+                learningGoals = lectureUnitRepository.findByIdWithLearningGoals(lectureUnit.getId()).map(LectureUnit::getLearningGoals).orElse(null);
+            }
+            else {
+                throw new IllegalArgumentException("Learning object must be either LectureUnit or Exercise");
+            }
```

Comment:
Since this is now repeated, you could move this as default method to the ILearningObject

Reply:
I'd rather not have so much service code in domain objects.